### PR TITLE
32-bit fixes

### PR DESCRIPTION
**Pull request by JornVernee**
- Fix 32-bit build errors and tests.
- Add negative test for 32-bit platform.
- Change CABI to fail more lazily when running on an unsupported platform, so that we can still initialize classes without getting an error.
- Change CLinker layouts to be null on unsupported platforms, instead of failing when initializing the class.

### CODE REVIEW

File: src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/CABI.java [hunk 1]
```diff
@@ -25,6 +25,8 @@
  */
 package jdk.internal.foreign;
 
+import static jdk.incubator.foreign.MemoryLayouts.ADDRESS;
+
 public enum CABI {
```
Comment (mcimadamore) [inline src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/CABI.java:32]: I think if you add an explicit UNSUPPORTED constant here, a lot of the downstream code will become more linear

File: src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java [hunk 2]
```diff
@@ -112,12 +113,25 @@
...
      * @throws IllegalAccessError if the runtime property {@code foreign.restricted} is not set to either
      * {@code permit}, {@code warn} or {@code debug} (the default value is set to {@code deny}).
+     * @throws UnsupportedOperationException if called when running on an unsupported platform.
      */
     static CLinker getInstance() {
         Utils.checkRestrictedAccess("CLinker.getInstance");
         return SharedUtils.getSystemLinker();
     }
 
+    /**
```
Comment (mcimadamore) [inline src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java]: Not sure about this - I wouldn't want to draw more attention on this than it deserves

File: src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java [hunk 3]
```diff
@@ -149,43 +163,43 @@ static CLinker getInstance() {
     MemorySegment upcallStub(MethodHandle target, FunctionDescriptor function);
 
     /**
-     * The layout for the {@code char} C type
+     * The layout for the {@code char} C type, or {@code null} on unsupported platforms.
```
Comment (mcimadamore) [inline src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java]: Here there are three main options:
1) the one you pursued
2) one where layout constants become instance fields - that's painful for clients because then static imports no longer work
3) turn static fields into static methods (and throw UOE)

All these options are not great IMHO. I think something like (1) might work - but we should consider some alternatives: for instance, we could just say that if a platform is unsupported, the contents of the layout constants is unspecified (once and for all, in the toplevel doc). Then, in the impl, we could either return null, or simply return e.g. SysV layouts, and call it done.

In other words, remove the isSupported predicate, move some of the javadoc in the toplevel doc where we say that on unsupported platforms (1) getInstance throws and (2) contents of layout constants is unspecified.

File: src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/CABI.java (hunk 1 above)
Comment (JornVernee) [reply src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/CABI.java:32]: Note really sure how this would help... It would mean having to add a default/UNSUPPORTED cases to switch expressions that switch over this enum. That's why I went with `null`.

Comment (mcimadamore) [reply src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/CABI.java:32]: Right - all clients which switch over current ABI will have to explicitly handle case where ABI is unsupported. Right now this is done by adding a null check before the switch, which is IMHO error-prone.

Comment (JornVernee) [reply src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/CABI.java:32]: The null check is only done once inside CABI::current though, not at every `switch`. Doesn't seem that bad? (i.e. the `null` is encapsulated in CABI)

Comment (JornVernee) [reply src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/CABI.java:32]: I'll push a commit that tries this out.

Comment (JornVernee) [reply src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/CABI.java:32]: After trying this out, I agree it does look nicer, since it's clear from a caller that an exception can occur. The throwing code can be shared in SharedUtils any ways.

File: src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java (hunk 2 above)
Comment (JornVernee) [reply src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java]: Removed now

File: src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java (hunk 3 above)
Comment (JornVernee) [reply src/jdk.incubator.foreign/share/classes/jdk/incubator/foreign/CLinker.java]: Solved by reverting to failing to initialize and added a note about that in the class javadoc.
